src/simperium/channel.js

```diff
@@ -102,42 +101,56 @@ internal.removeAndSend = function( id ) {
 	return this.store.get( id ).then( remove );
 };
 
-// We've receive a full object from the network. Update the local instance and
-// notify of the new object version
+/**
+ * Updates the ghost store with the updated ghost data based on the data that
+ * comes from applying the patch to the original.
+ *
+ * @param { string } id - the bucket object key for the object being updated
+ * @param { number } version - the version number the ghost is being updated to
+ * @param { Object } data - the new data for the ghost for this version
+ * @param { Object } original - the original data for the ghost before patch is applied
+ * @param { Object } patch - the patch applied to the original to get the resulting data
+ * @param { Object } [acknowledged ] - the change sent by this client matching the received network change
+ * @returns { Promise<*> } resolves once the ghost has been updated
+ */
 internal.updateObjectVersion = function( id, version, data, original, patch, acknowledged ) {
-	var notify,
-		changes,
-		change,
-		patch,
-		localModifications,
-		remoteModifications,
-		transformed,
-		update;
+	const save = () => this.store.put( id, version, data );
+
+	if ( acknowledged ) {
+		return save().then( () => {
+			internal.updateAcknowledged.call( this, acknowledged );
+		} );
+	}
 	// If it's not an ack, it's a change initiated on a different client
 	// we need to provide a way for the current client to respond to
 	// a potential conflict if it has modifications that have not been synced
-	if ( !acknowledged ) {
-		changes = this.localQueue.dequeueChangesFor( id );
-		localModifications = change_util.compressChanges( changes, original );
-		remoteModifications = patch;
-		transformed = change_util.transform( localModifications, remoteModifications, original );
-		update = data;
-
-		// apply the transformed patch and emit the update
-		if ( transformed ) {
-			patch = transformed;
-			update = change_util.apply( transformed, data );
-			// queue up the new change
-			change = change_util.modify( id, version, patch );
-			this.localQueue.queue( change );
-		}
-
-		notify = this.emit.bind( this, 'update', id, update, original, patch, this.isIndexing );
-	} else {
-		notify = internal.updateAcknowledged.bind( this, acknowledged );
-	}
+	return this.onBeforeNetworkChange( id, data, original, patch )
+		.then( ( local ) => {
+			// pull all pending changes in the local queue for the object of id
+			const localModifications = change_util.diff( original, local ),
+				// rebases the local changes on top of the network patch
+				transformed = change_util.transform( localModifications, patch, original );
+
+			// remove pending changes
+			this.localQueue.dequeueChangesFor( id );
```

Comment:
are we saying here that _if we have changes in Simperium locally that are waiting to be sent to the server_ and also that _the client application has its own changes_ then we can effectively forget about the ones we _were going to send_ and start from scratch based on the updates from the client?

that would make sense and explain why we're ignoring the changes from this line

Reply:
Yes. If the client application tells us "this the data I have for this object", then whatever was in the queue no longer matters.

Reply:
sounds good. we don't get a history revision for that data but it won't matter